[MISC][Bugfix] Use less CPU when message queue has been empty for some time

p12tic opened the pull request:
In setups which have long inactivity periods it is desirable to reduce system power consumption when vllm does nothing. This would lead to more CPU thermal headroom when a request eventually comes, especially when multiple GPUs are connected as each GPU would otherwise pin one thread at 100% CPU usage.

~I didn't include any configuration knobs because I couldn't think of anyone who would adversely impacted by this change. It introduces a maximum additional latency of 100ms when vllm workers are inactive for 10 seconds or more. Seems like someone who has inactive vllm wouldn't care about small additional latency. But please let me know if you think otherwise.~

EDIT: Final version of the PR contains new environment variable VLLM_SLEEP_WHEN_IDLE to enable CPU usage reduction. Set VLLM_SLEEP_WHEN_IDLE=1 to reduce power consumption at the expense of small amount of additional latency.

FIX #14799, https://github.com/vllm-project/vllm/issues/16968, https://github.com/vllm-project/vllm/issues/16660

MattJCroteau commented:
I was experiencing 100% CPU load on 0.8.3, applied this PR and it's now fixed. Thank you.

zacksiri commented:
Any chance this can get merged?

p12tic commented:
Rebased on latest `main`

p12tic commented:
@youkaichao, @WoosukKwon, @mgoin Could you please take a look? I cannot request reviews, thus pinging directly. Thank you very much.

p12tic commented:
Not sure why tests are not working, seems like instability. First iteration of the PR had green tests.

ahmeteminkocal commented:
+1

nOneKzero commented:
+1

lingej commented:
+1

bengsV commented:
+1

Jollokim commented:
+1

xanadu-3g commented:
@youkaichao Consider our use case: the vllm server gets a few thousand queries a day, mostly in working hours, the average inference takes 5 seconds, a few millisecond more latency has no relevance. Starting vllm takes more than a minute, so shutting it down is not an option. Saturating 2 cpu cores increases the system's power usage by 2,5x

p12tic commented:
> Could you change the behavior so it is off by default, preserving original behavior?

Will do, thanks.

njhill commented:
> @njhill
> 
> > I was wondering whether we can just use zmq instead of the shm spin there?
> 
> Using zmq does not necessarily mean that there won't be busy loop. I fixed very similar issue in sglang in their zmq code path ([sgl-project/sglang#6026](https://github.com/sgl-project/sglang/pull/6026)).

@p12tic there won't be any busy loop if zmq is used properly. Maybe they were reading nonblocking in a busy loop but that isn't what we do anywhere. I think we should just switch to that unless there's a measurable performance difference to using shm.

p12tic committed:
[MISC] Use less CPU when message queue has been empty for some time

In setups which have long inactivity periods it is desirable to reduce
system power consumption when vllm does nothing. The simpliest is to
reduce polling frequency when there is no activity for a certain period
of time.

Fixes: https://github.com/vllm-project/vllm/issues/14799
Signed-off-by: Povilas Kanapickas <povilas@radix.lt>

p12tic commented:
@mgoin @youkaichao I have implemented `--sleep-on-idle` CLI option, tested locally, CI tests are green. The PR is ready for review again. Thanks!

youkaichao commented:
> @mgoin @youkaichao I have implemented `--sleep-on-idle` CLI option, tested locally, CI tests are green as well. The PR is ready for review again. Thanks!

I don't think it's worthwhile to add cli arg, since i suspect this is a less frequent knob. I think maybe an env var would be enough.

> I was wondering whether we can just use zmq instead of the shm spin there?j (it already supports both and falls back to zmq for large msg or remote cases anyhow). I suspect the perf difference may not be noticeable.

I'm not sure if zmq would be competitive here. The shm implementation leverages the fact that we have fixed number of readers and only one writer. The zmq PUB-SUB socket cannot assume fixed number of readers.

p12tic commented:
@youkaichao 

> I don't think it's worthwhile to add cli arg, since i suspect this is a less frequent knob. I think maybe an env var would be enough.

I will change the implementation according to your suggestion, but I think it's worth considering that this PR is actually rather popular. If counting by 👍 reactions it's currently the second open PR and within top 0.5% of all vllm PRs ever opened.

The option will be relevant for all residential and small commercial deployments. Running couple cores at 100% increases power usage and cooling requirements substantially due to turbo boost. On many CPUs running 1/8 total threads (1/4 cores) at 100% will use something like 4/5 max TDP of power.

I understand that a single vllm deployment by one of the large players is more GPUs than all small deployments combined. However in terms of how many people use vllm, I would argue that there is a large number of relatively small deployments, so it makes sense to save people time by exposing the knob somewhere visible instead of hiding it behind an environment variable.

MattJCroteau commented:
I would humbly add to the comment @p12tic is making.  Our vllm deployment is indeed small and it's just one of the things the server is running. It's eating 100% CPU just sitting 'idle'.  A tight polling loop that consumes 100% CPU is a common bug in software. IMHO this PR should have merged several months ago and been met with a 'Exceelent bug fix, Thanks!'

p12tic commented:
@youkaichao Just a friendly ping regarding my comment: https://github.com/vllm-project/vllm/pull/16226#issuecomment-2905501702

mgoin commented on vllm/config.py:
```diff
@@ -1741,6 +1741,9 @@ class is dynamically inherited by the worker class. This is used to inject
     """Global rank in distributed setup."""
 
+    sleep_on_idle: bool = False
+    """Reduce CPU usage when sglang is idle."""
```
```suggestion
    """Reduce CPU usage when vLLM is idle."""
```

p12tic replied:
Fixed now, thanks. I should have reread more carefully.

mgoin commented on vllm/distributed/device_communicators/shm_broadcast.py:
```diff
@@ -43,6 +43,43 @@ def sched_yield():
+        sched_yield()
+
+
+class SpinSleepTimer(SpinTimer):
+    """
+    In setups which have long inactivity periods it is desirable to reduce
+    system power consumption when vllm does nothing. This would lead to more
+    CPU thermal headroom when a request eventually comes, especially when
+    multiple GPUs are connected as each GPU would otherwise pin one thread at
+    100% CPU usage.
+
+    The simplest solution is to reduce polling frequency when there is no
+    activity for a certain period of time.
+    """
+
+    def __init__(self, busy_loop_s: float = 10.0, wait_sleep_s: float = 0.1):
+        self.last_activity = time.monotonic()
+        self.busy_loop_s = busy_loop_s
+        self.wait_sleep_s = wait_sleep_s
+
+    def record_activity(self):
+        self.last_activity = time.monotonic()
+
+    def spin(self):
+        curr_time = time.monotonic()
+        if curr_time >= self.last_activity + self.busy_loop_s:
+            time.sleep(self.wait_sleep_s)
+        else:
+            sched_yield()
```
Should this be using zmq now per the discussion?

p12tic replied:
Thanks for review. I believe this would be a nontrivial change with way higher risks than what has been done now, especially with regards to performance. zmq would be the ideal solution, but this PR will already gain 90-99% of potential savings, so without other reasons zmq is not worth switching to.

p12tic commented:
I reduced the idle time until sleeping turns on from 10 seconds to 3 seconds. Since this feature is now controlled by an option which is not default, it's possible to be a little bit more aggressive with power savings.

njhill reviewed (commented):
Thanks @p12tic, this looks good to me and I agree it's reasonable to add and separately evaluate eliminating the shm spin altogether.

I feel like this would be better as an env var rather than a config / cli arg though, especially since we may remove it in the near future. See https://github.com/vllm-project/vllm/blob/main/vllm/envs.py.

p12tic commented:
> @p12tic are you ok with updating this to use an env var instead of CLI arg?

Sorry, I only added reaction to your comment.

p12tic committed:
[MISC] Add VLLM_SLEEP_WHEN_IDLE env arg

Signed-off-by: Povilas Kanapickas <povilas@radix.lt>

p12tic commented:
@njhill Sorry for the delay again, I've implemented and tested env variable based solution.

njhill reviewed (approved):
Thanks @p12tic

njhill commented:
Actually @p12tic would you be ok to add a test for this? Can just be copy of one of the existing simple tests but with the env var set, or else we're not exercising that path at all.

zacksiri commented:
@p12tic thank you so much for this patch. I've been following this thread, your patience and persistence is greatly appreciated.

Thanks to the vllm team for merging this patch, I believe it will beneficial for a lot of users who have smaller deployments.

njhill reviewed (approved):
Thanks again @p12tic for your work and patience!
